design-book/src/roadmap.md

```diff
@@ -0,0 +1,100 @@
+This list is not "all of the features that the editor will have": other stages will follow as these features are added.
+It's best not to plan too far in advance!
+
+## Stage 0: Hello Project
+
+- [x] make a repo
+- [x] add a README
+- [x] add a CONTRIBUTING.md
+- [x] add a basic Bevy project
+- [x] define the broad plan for what we're doing
+- [x] define basic constraints for how we're building the editor
+- [x] define a shared vision for what the editor is (and is not)
+
+## Stage 1: Technically an Editor
+
+- [ ] can load scenes from disk using a native file picking widget
+- [ ] can display scenes in a viewer
+  - [ ] 2D
+  - [ ] 3D
+- [ ] simple camera controller
+  - [ ] 2D
+  - [ ] 3D
+- [ ] gizmos can be toggled
+  - [ ] UI outlines
+  - [ ] lights
+  - [ ] AABBs
+- [ ] the list of entities and their components is shown in the inspector
+  - [ ] supports hierarchy via a folding tree view
+- [ ] existing component values can be modified using a graphical interface
+- [ ] resources can be inspected, showing their values and allowing modification
+- [ ] loaded assets can be inspected, providing basic information about them
+- [ ] scenes can be saved back to disk
+
+## Stage 2: Editor-Game Interaction
+
+- [ ] users can press a button in the editor, launching the game from the very beginning
+- [ ] users can press a button in the editor, and their game will run, loading the currently active scene
+- [ ] components can be added and removed, including components specific to the user's game
+- [ ] entities can be spawned and despawned via a GUI
+- [ ] entity hierarchies can be spawned and despawned via a GUI
+  - [ ] manually
+  - [ ] by composing scene files
+- [ ] game-specific rendering techniques display correctly in the editor's scene editing
+- [ ] distinct editor and game views that can be run simultaneously
+
+## Stage 3: UI/UX Foundations
+
+- [ ] tooltips
+- [ ] hotkeys
+- [ ] primitive widgets
+  - [ ] text entry
+  - [ ] numeric entry
+  - [ ] radio buttons
+  - [ ] dropdown
+  - [ ] checkboxes
+- [ ] editor UI foundations
+  - [ ] resizable panes
+  - [ ] scrollable tool detail panels
+  - [ ] toolbar
+  - [ ] Blender-style workspaces
+  - [ ] context menus
+  - [ ] command palette
+- [ ] dedicated hotkeys and controls for the most common operations:
+  - [ ] manipulating camera
+  - [ ] modifying transforms
+  - [ ] toggling visibility
+  - [ ] adding and removing entities
+  - [ ] saving the project
+  - [ ] opening the command palette
+
+## Stage 4: Fundamental Features
+
+- [ ] undo-redo abstraction
+- [ ] preferences
+  - [ ] hotkey rebinding
+  - [ ] camera controls
+  - [ ] preferred language
+- [ ] scene object picking: click on objects in the scene to select them in the inspector
+- [ ] entities can be looked up by name in the inspector
+- [ ] system stepping support
+- [ ] create new Bevy project
+  - [ ] blank
+  - [ ] from template
+- [ ] localization framework (first-party: English-only)
```

Comment:
Suggest to add:
- Extensions (allow to create new panels and tools in editor by game or downloadable editor extenstions without editor recompiling)  (may be stage 5?)
- FPS counter
- Clone entity with all components/clone entity hieararchy
- Drag and drop for assets/scene files 
- Asset folder inspector (it will be very very very useful to create Unity-like project inspector with preview icons for materials/scenes/images/meshes/any another custom files which can generate preview and clickable)

Reply:
According to a Unity dev blog post I once read, runtime extensions to the editor are the most valued thing in their Asset store. I'd say this is pretty important to keep in the back of our mind. Sure, no use to define an extension API in a prototype, agreed, but since this document is intended to go beyond the prototype, I'd mention it.

Reply:
That view kind of per-supposes an asset store doesn't it?

Reply:
> I'd punt on runtime extensions entirely for now.

I view this document not as something we need to polish to perfection, but as an approximate path towards a useful editor for us. Bevy is very strong in its ability to be expanded by community plugins and its large community that writes these plugins. To be useful to the Bevy community, the editor should also be easily expandable by community plugins. Therefore, I want to propose planning a moment to start discussions on this matter.

Reply:
> That view kind of per-supposes an asset store doesn't it?

AFAIK our assets page is supposed to become something like that at some point in the future, yes. Correct me if I'm wrong :)
But even if not, our current assets page could easily host free and open source editor extensions today if the editor existed, so I don't see a problem there.

Reply:
Ok let's talk about this. I guess, if we are designing the editor primarily for teams, and we are assuming (for now) everyone on the team is either building the editor from source or getting binaries from other team-members, who are _runtime_ extensions for?

Reply:
> But even if not, our current assets page could easily host free and open source editor extensions today if the editor existed, so I don't see a problem there.

The difference is that, if the source is available, you can compile it yourself. Stores (and runtime plugins) are generally for redistributing proprietary & closed-source code.

Reply:
> Ok let's talk about this. I guess, if we are designing the editor primarily for teams, and we are assuming (for now) everyone on the team is either building the editor from source or getting binaries from other team-members, who are runtime extensions for?
> 
> If we get to a point where app code and editor code are not directly linked, I could see the value maybe. But shouldn't we focus on getting to that point first?

From thie design-constraints.md
> The editor should be distributed as an executable, either directly from the Bevy website or via Steam.

Our final goal is to create a distributable exe. To make it extensible with plugins at some point, we will need to add runtime extensions. However, I'm not opposed to the idea that until that moment, the editor can be extended using code extensions, but the transition point is worth discussing.

Reply:
> If we get to a point where app code and editor code are not directly linked, I could see the value maybe.

It's already supposed to be on stage 2

Reply:
I don't really want to nail down any specifics, it just seems important to note that *some day*, a level designer should have *some way* of receiving an extension that will make their work easier. I don't see "the artist has to ask the local programmer to compile them a custom editor" as a particularly friendly solution. It's fine as an intermediate, but it shouldn't be the end goal.
Agreed that this is not something we don't need to support from the get-go. But it seems like something of such great value that it would be quite the omission to not mention it *somewhere*.

Reply:
> A level designer should have some way of receiving an extension that will make their work easier. I don't see "the artist has to ask the local programmer to compile them a custom editor" as a particularly friendly solution.

I get this argument, but this is speculating. I really would rather we build something usable and then get actual user feedback. At this stage it really feels to me like a common solution looking for a problem to solve. In conversations I've had with some of bevy's existing private teams, none of them have actually flagged runtime plugins as a priority.

Reply:
There's been a significant amount of architecture discussion in `#editor-dev` on discord, and it has answered a bunch of these questions. I'd really like if we could maybe condense that down into a `architecture.md` document as well. We can update it as we progress through the stages.

To start with, it seems like the editor will be a bevy plugin which is statically bundled into the app. That's just the simplest way to get started.

Reply:
> > A level designer should have some way of receiving an extension that will make their work easier. I don't see "the artist has to ask the local programmer to compile them a custom editor" as a particularly friendly solution.
> 
> I get this argument, but this is speculating. I really would rather we build something usable and then get actual user feedback. At this stage it really feels to me like a common solution looking for a problem to solve. In conversations I've had with some of bevy's existing private teams, none of them have actually flagged runtime plugins as a priority.

In my opinion, no one prioritizes runtime plugins as something that needs to be done from the start. On the contrary, it's suggested to define everything that needs to be done first. And at the end of the list, write "and here we'll deal with dynamic linking and plugins."

Reply:
> And at the end of the list, write "and here we'll start deal with dynamic linking, plugins and hot reload." At the very least, I simply don't want this important thing to be left out of long-term planning.

Exactly my position.